libs/test/MobileSyncTest/src/com/salesforce/androidsdk/mobilesync/target/SyncUpTargetTest.java

```diff
@@ -197,6 +197,86 @@ else if (name.equals("")) {
         idToFields.putAll(idToFieldsGoodNames);
     }
 
+    /**
+     * Create a few records - some with no sobject type
+     * Sync up
+     * Make sure the records with no sobject type fail to sync and that last error reflects the problem
+     * @throws Exception
+     */
+    @Test
+    public void testSyncUpWithNoType() throws Exception {
```

Comment:
New tests:
- no type
- bad typ
Those tests run for both SyncUpTarget and BatchSyncUpTarget.
Before the code change, the test would repro the infinite looping.